syntax = "proto3";

message Country {
  string name = 1;
}

message SubCountry {
  string name = 1;
}

message City {
  string name = 1;
}

message Ip {
  string direction = 1;
}

message Empty {
}

service GeoService {
  rpc GetAllCountries(Empty) returns (GetAllCountriesReply);
  rpc GetSubCountries(Country) returns (GetSubCountriesReply);
  rpc GetCities(SubCountry) returns (GetCitiesReply);
  rpc GetLocationOfIp(Ip) returns (GetLocationOfIpReply);
}

message GetAllCountriesReply {
  repeated string countries = 1;
}

message GetSubCountriesReply {
  repeated string subCountries = 1;
}

message GetCitiesReply {
  repeated string cities = 1;
}

message GetLocationOfIpReply {
   optional string country = 1;
   optional string state = 2;
   optional string error = 3;
}